File: src/homepod-base/driver.ts
import { Driver } from '@basmilius/homey-common';
import type { AppleApp } from '../types';
import type Homey from 'homey';
import HomePodBasePairing from './pairing';

export default abstract class HomePodBaseDriver extends Driver<AppleApp> {
    abstract get modelFilter(): RegExp;

    async onPair(session: Homey.Driver.PairSession): Promise<void> {
        const pairing = new HomePodBasePairing(session, this.discovery.getStrategy('airplay'), this.modelFilter, this.getDevices());

        pairing.on('error', err => {
            // todo: Show error screen or something.
            this.error(err);
        });

        pairing.on('log', log => {
            this.log(log);
        });

        await pairing.start();
    }
}

File: src/homepod_base/driver.py
from __future__ import annotations

import re
from abc import abstractmethod
from typing import Any

from homey.driver import Driver

from .pairing import HomePodBasePairing


class HomePodBaseDriver(Driver):
    """Abstract base driver for HomePod and HomePod Mini."""

    @property
    @abstractmethod
    def model_filter(self) -> re.Pattern:
        """Regex pattern to match device model strings during pairing."""

    async def on_pair(self, session: Any) -> None:
        pairing = HomePodBasePairing(
            session=session,
            strategy=self.get_discovery_strategy(),
            model_filter=self.model_filter,
            known_devices=list(self.get_devices()),
        )

        pairing.on_log = lambda msg: self.log(msg)
        pairing.on_error = lambda err: self.error(err)

        await pairing.start()
